augur/clades.py

```diff
@@ -147,16 +144,40 @@ def assign_clades(clade_designations, all_muts, tree, ref=None):
         sorted_nodes = sorted(node_counts, key=lambda x: x.leaf_count, reverse=True)
         if len(sorted_nodes) > 0:
             target_node = sorted_nodes[0]
-            clade_membership[target_node.name] = {'clade_annotation': clade_name, 'clade_membership': clade_name}
+            basal_clade_nodes[target_node.name] = clade_name
+            clade_membership[target_node.name] = clade_name # basal nodes are members of the clade
 
     # third pass to propagate 'clade_membership'
     # don't propagate if encountering 'clade_annotation'
     for node in tree.find_clades(order = 'preorder'):
         for child in node:
-            if 'clade_annotation' not in clade_membership[child.name]:
-                clade_membership[child.name]['clade_membership'] = clade_membership[node.name]['clade_membership']
+            # if the child doesn't define the start of its own clade, but the parent belongs to a clade, then inherit that membership
+            if child.name not in basal_clade_nodes and node.name in clade_membership:
+                clade_membership[child.name] = clade_membership[node.name]
+    return (basal_clade_nodes, clade_membership)
+
 
-    return clade_membership
+def create_node_data_structure(basal_clade_nodes, clade_membership, args):
+    node_data = {}
+
+    if (not args.label_name and not args.trait_name):
```

Comment:
I wanted to allow workflows to continue without needing changes. There were 2 ways I could think of ding this:

1. `augur clades` without these 2 arguments used the old behaviour & exported both `clade_membership` and `clade_annotation` as node traits. These would be picked up by `augur export v2` and the latter turned into the branch label `clade. The downside is that the file structure for `augur clades` is different if you don't provide arguments than if you do.
2. augur clades` without these 2 arguments now stores clade membership as before but stores branch labels in the new `branch_labels` structure under a key `clade`. This structure needs no special interpretation by `augur export v2`, and we will end up with identical Auspice JSONs as previously. The downside is that the format of the file produced by `augur clades` differs. 

I went with option 2, but am open to other suggestions.

Reply:
Can we make `clade` and `clade_membership` the default values for the label and attribute names and store their values in the new structure? Is there a reason to make these required arguments in the future?

As a user, I would be surprised to find I need to define these values when I've never needed to before and I'd probably just use the defaults anyway.

If we allow these arguments to have default values, then we only need five lines of this function and those can be moved into `run`.

Reply:
Good question! My reason for requiring users to specify is to allow users to call `augur clades` multiple times in a single workflow, sometimes with branch labels, sometimes without (and perhaps sometimes without trait names etc). If these had defaults, then the defaults will end up exported in the auspice JSON which may be undesired and potentially confusing as it wouldn't be clear which invocation of `augur clades` produced it. 

Concrete examples for discussion:

```
augur clades ... --trait-label pango # no branch label - not guaranteed monophyletic
augur export ...
```
This is going to end up with "clade" branch labels representing pangolin clades, which wasn't the desired intention of the workflow.

```sh
augur clades ... --trait-label pango # no branch label - not guaranteed monophyletic
augur clades ... --branch-label emerging_lineage # no trait labels
augur clades ... --trait-label WHO
augur export ...
```
We're going to get branch labels "clade", which I _think_ will be WHO clades (this is an implementation detail of `augur export` as to which one is picked - worst case they may be a mixture!). We're also going to get a colouring `clade_membership`, which actually refers to emerging lineage, but it isn't obvious why.

Reply:
Got it. Those examples helped! If I understand correctly, there are two separate issues that we’re trying to address by requiring these new arguments:

1. Users can run `augur clades` multiple times with the same (default) attribute/label names and `augur export` will happily consume these and prefer one over the other in a surprising or unpredictable order (at least for the user).

2. Users can run `augur clades` with one or both of the new arguments depending on what they want to annotate (clade attributes only, branch labels only, or both). Using default values would produce unwanted outcomes when users choose only attributes or only branch labels and also gets an annotation for the other possible representation.

Is that summary generally correct?

I can see how requiring these arguments tries to protect against conflicting node/branch attributes in `augur export`. This is similar to why `augur distance` requires `--attribute-name`. But this seems to be a general problem with the export logic where we don't check (I think?) for collisions in attribute names from different data sources. So, even though we require the user to specify attribute names, there is no reason they couldn’t specify the same names in separate commands and still get a surprise collision. Another way to address issue 1 would be to check for these types of collisions in `augur export` and either warn the user or throw an error. In addition to addressing Issue 1 here, this solution would also address other cases in the real world where people accidentally define the same attribute in separate runs of other augur commands. If issue 1 was the only issue, I'd still prefer to set sane defaults and not expect the user to change their behavior.

Issue 2 is one I missed on my initial read through the code (that you can define attributes or labels and not both). Still, I wonder about how bad it would be for users to get branch labels when they only request attributes. If I ask for emerging lineage branch labels and I get an emerging lineage color-by as a side effect, is that a bug or a feature? Is the worst case scenario here that the user is annoyed to get an annotation they don’t expect? We have already been providing these dual annotations, so would they actually be surprised? The main issue seems to be when the default names for the other representation conflict across multiple runs of the same command.

This example also makes me wonder about the value of using different names for attributes and branch labels. The name we use describes the data source of the clade annotations and not how Auspice represents clade annotations. That a clade appears as a color-by or branch label in Auspice is a separate technical consideration.

I also don’t see the harm in annotating both node and branch attributes with the same name by default. What if use the same name for both attributes and keep a sane default value (e.g., “clade”)? This approach allows the user who only runs clades once in a workflow to change nothing and run:

```bash
# Annotate both node and branch attributes. The user gets
# output that differs in its JSON structure but appears the
# same way in Auspice as it always has. Augur export knows
# how to handle the new JSON structure in this same release
# of Augur, so we don't need any special checks for backward
# compatibility.
augur clades \
    --clades clades.tsv \
    --output clades.json
```

Then, the user who wants to run multiple instances of clades in a single workflow can run the following commands to be more explicit about their attribute names:

```bash
# Provide explicit node/branch attribute names.
augur clades \
    --clades clades.tsv \
    --attribute-name nextstrain_clade \
    --output clades.json

augur clades \
    --clades pango.tsv \
    --attribute-name pango \
    --output pango.json
```

If users specify the same attribute name in separate data sources, `augur export` should complain loudly:

```bash
# Use the default attribute name. Annotate both node and
# branch attributes.
augur clades \
    --clades clades.tsv \
    --output clades.json

# Accidentally reuse the same default attribute name.
augur clades \
    --clades pango.tsv \
    --output pango.json

# Validate attribute names from distinct data sources.
augur export v2 \
    --node-data clades.json pango.json
    ...
ERROR: Multiple node data files ("clades.json", "pango.json") provide the same attribute name ("clade"). Resolve conflicting attribute names (e.g., by specifying `--attribute-name`) for these data files and try again.
```

Allowing default values makes this a backward-compatible change where most users do not have to do anything. Using the same name for node and branch attributes allows the user to know which `augur clades` invocation produced those attributes and not have to think about how the clade annotation is represented in Auspice. Checking for collisions in node/branch attribute names in `augur export` alerts the user when they accidentally reuse the same attribute names in separate invocations and tells them how to correct the problem (and fixes a more general issue with `augur export`). I think this approach also simplifies the code in this PR a bit by reusing the same attribute name.

What do you think?

Reply:
Considering the scope of `augur clades`, I agree with these points (branch labels and node attrs stored under the same attribute name, both exported, one optional `--attribute-name" arg with a default of "clade"), and am happy to make the changes to that code.

Where it gets tricky is in `augur export`, because that is when we combine various pieces of data into a desired visualisation. To cleanly demarcate data generation vs visualisation, I do think these complexities are the remit of `augur export`. How we determine what's exported has always been somewhat poorly documented and without looking at the code I can't remember what happens in many cases:
* What happens if pieces of (meta)data differ in the metadata TSV and a node-data file?
* Is a node-data attribute always exported as a colouring, even if we provide a list of desired colourings in an auspice config JSON which doesn't specify it?
* what about if we provide a list of colourings on the command line?
* What about if we do both?
* Are there special cases? (Yes, [at least 18](https://github.com/nextstrain/augur/blob/15dfc8bc5c5a570066ddd5601c9aa82e0c1d2b7d/augur/export_v2.py#L667-L687) and probably more.)

This relates to this PR as I think we want to have answers to the following questions:
* Previously, clade colourings were exported as "clade_membership", and this was _always_ set as a colouring if a node-data file provided it. This is easy to update to "clade" if we we want to keep this behaviour.
* If an auspice config JSON specified a colouring for `key="clade_membership"`, which many do, but such an attribute is no longer provided in any node-data JSONs, what do we do?
* Is there a way to limit the exported colourings, i.e. is specifying a list of colourings in the config JSON able to prevent the export of a node-data attribute?
* Currently there's no general way to export branch labels (that's part of this PR). Do we extend the auspice config PR to allow these to be specified? Does this act the same way as `colorings`?